pkg/validate/container_linux.go

```diff
@@ -307,3 +307,199 @@ func createOOMKilledContainer(
+
+		AfterEach(func() {
+			By("stop PodSandbox")
+			rc.StopPodSandbox(context.TODO(), podID)
+			By("delete PodSandbox")
+			rc.RemovePodSandbox(context.TODO(), podID)
+		})
+
+		testRRO := func(rc internalapi.RuntimeService, ic internalapi.ImageManagerService, rro bool) {
+			if rro && !runtimeSupportsRRO(rc, "") {
+				Skip("runtime does not implement recursive readonly mounts")
+				return
+			}
+
+			By("create host path")
+			hostPath, clearHostPath := createHostPathForRROMount(podID)
+			defer clearHostPath() // clean up the TempDir
+
+			By("create container with volume")
+			containerID := createRROMountContainer(rc, ic, podID, podConfig, hostPath, "/mnt", rro)
+
+			By("test start container with volume")
+			testStartContainer(rc, containerID)
+
+			By("check whether `touch /mnt/tmpfs/file` succeeds")
+			command := []string{"touch", "/mnt/tmpfs/file"}
+			if rro {
+				command = []string{"sh", "-c", `touch /mnt/tmpfs/foo 2>&1 | grep -q "Read-only file system"`}
+			}
+			execSyncContainer(rc, containerID, command)
+		}
+
+		It("should support non-recursive readonly mounts", func() {
+			testRRO(rc, ic, false)
+		})
+		It("should support recursive readonly mounts", func() {
+			testRRO(rc, ic, true)
+		})
+		testRROInvalidPropagation := func(prop runtimeapi.MountPropagation) {
+			if !runtimeSupportsRRO(rc, "") {
+				Skip("runtime does not implement recursive readonly mounts")
+				return
+			}
+			hostPath, clearHostPath := createHostPathForRROMount(podID)
+			defer clearHostPath() // clean up the TempDir
+			mounts := []*runtimeapi.Mount{
+				{
+					HostPath:          hostPath,
+					ContainerPath:     "/mnt",
+					Readonly:          true,
+					RecursiveReadOnly: true,
+					SelinuxRelabel:    true,
+					Propagation:       prop,
+				},
+			}
+			const expectErr = true
+			createMountContainer(rc, ic, podID, podConfig, mounts, expectErr)
+		}
+		It("should reject a recursive readonly mount with PROPAGATION_HOST_TO_CONTAINER", func() {
+			testRROInvalidPropagation(runtimeapi.MountPropagation_PROPAGATION_HOST_TO_CONTAINER)
+		})
+		It("should reject a recursive readonly mount with PROPAGATION_BIDIRECTIONAL", func() {
+			testRROInvalidPropagation(runtimeapi.MountPropagation_PROPAGATION_BIDIRECTIONAL)
+		})
+		It("should reject a recursive readonly mount with ReadOnly: false", func() {
+			if !runtimeSupportsRRO(rc, "") {
+				Skip("runtime does not implement recursive readonly mounts")
+				return
+			}
+			hostPath, clearHostPath := createHostPathForRROMount(podID)
+			defer clearHostPath() // clean up the TempDir
+			mounts := []*runtimeapi.Mount{
+				{
+					HostPath:          hostPath,
+					ContainerPath:     "/mnt",
+					Readonly:          false,
+					RecursiveReadOnly: true,
+					SelinuxRelabel:    true,
+				},
+			}
+			const expectErr = true
+			createMountContainer(rc, ic, podID, podConfig, mounts, expectErr)
+		})
+	})
+})
+
+func runtimeSupportsRRO(rc internalapi.RuntimeService, runtimeHandlerName string) bool {
```

Comment:
If `framework.TestContext.RuntimeHandler` is set then its value should have the highest priority.

Reply:
Fixed to check `framework.TestContext.RuntimeHandler`